demos/audiovisual/src/audiovisualapp.cpp

```diff
@@ -0,0 +1,240 @@
+		if (!error.check(mRenderCameraEntity != nullptr, "unable to find entity with name: %s", "RenderCameraEntity"))
+			return false;
+
+
+		mRenderEntity = mScene->findEntity("RenderEntity");
+		if (!error.check(mRenderEntity != nullptr, "unable to find entity with name: %s", "RenderEntity"))
+			return false;
+
+		mWorldEntity = mScene->findEntity("WorldEntity");
+		if (!error.check(mWorldEntity != nullptr, "unable to find entity with name: %s", "WorldEntity"))
+			return false;
+
+		// Cache mask
+		mLitRenderMask = mRenderService->getRenderMask("Lit");
+
+		// Reset first frame flag on resource reload
+		mResourceManager->mPostResourcesLoadedSignal.connect(mReloadSlot);
+
+		// All done!
+		return true;
+	}
+	
+	
+	// Update app
+	void audiovisualApp::update(double deltaTime)
+	{
+		// Use a default input router to forward input events (recursively) to all input components in the default scene
+		nap::DefaultInputRouter input_router(true);
+		mInputService->processWindowEvents(*mRenderWindow, input_router, { &mScene->getRootEntity() });
+
+		// GUI
+		ImGui::Begin("Controls");
+		ImGui::Text(getCurrentDateTime().toString().c_str());
+		ImGui::Text(utility::stringFormat("Framerate: %.02f", getCore().getFramerate()).c_str());
+		ImGui::Text(utility::stringFormat("Frametime: %.02fms", deltaTime * 1000.0).c_str());
+
+		for (const auto& gui : mResourceManager->getObjects<ParameterGUI>())
+			gui->show(false);
+
+		ImGui::End();
+	}
+	
+	
+	// Render app
+	void audiovisualApp::render()
+	{
+		// Signal the beginning of a new frame, allowing it to be recorded.
+		// The system might wait until all commands that were previously associated with the new frame have been processed on the GPU.
+		// Multiple frames are in flight at the same time, but if the graphics load is heavy the system might wait here to ensure resources are available.
+		mRenderService->beginFrame();
+
+		// Compute
+		std::vector<ComputeComponentInstance*> compute_comps;
+		mWorldEntity->getComponentsOfTypeRecursive<ComputeComponentInstance>(compute_comps);
+		if (mRenderService->beginComputeRecording())
+		{
+			mRenderService->computeObjects(compute_comps);
+			mRenderService->endComputeRecording();
+		}
+
+		std::vector<RenderableComponentInstance*> render_comps;
+		mWorldEntity->getComponentsOfTypeRecursive<RenderableComponentInstance>(render_comps);
+
+		// Headless
+		if (mRenderService->beginHeadlessRecording())
+		{
+			if (mFirstFrame)
+			{
+				auto cube_targets = mResourceManager->getObjects<CubeRenderTarget>();
+				if (!cube_targets.empty())
+				{
+					assert(cube_targets.front() != nullptr);
+					auto target = cube_targets.front();
+
+					for (auto* comp : render_comps)
+					{
+						if (comp->get_type() != RTTI_OF(RenderableMeshComponentInstance))
+							continue;
+
+						auto* resource = comp->getComponent<RenderableMeshComponent>();
+						if (resource == nullptr)
+							continue;
+
+						if (resource->mID == "RenderStars")
```

Comment:
I thought we had something in place for rendering cube-maps on initialization, why this verbose render step on frame 1 here? Also: `getObjects()` should not be used, it is considered bad practice because it could return recently replaced (not active) targets, which in this case is extra bad because we're only doing it on frame one?